File: services/servico-order/src/domain/events/DomainEvent.ts
/**
 * Contrato genérico de evento de domínio publicado pelo ServicoOrder,
 * alinhado com a especificação de eventos da Tabela "Especificação final
 * dos microsserviços do FTGO" (EC-M4.3).
 *
 * Em produção, cada evento seria publicado em um tópico Kafka
 * (ex.: `order_events`), conforme definido na EC-M5.2. Nesta primeira
 * versão (apenas o ServicoOrder), a publicação ocorre em um EventBus
 * em memória (ver src/infrastructure/events/InMemoryEventBus.ts),
 * substituível por um adaptador Kafka real quando os demais
 * microsserviços (Accounting, Kitchen, Delivery, Restaurant, Consumer)
 * forem implementados.
 */
export interface DomainEvent<TPayload = unknown> {
  /** Nome do evento (ex.: "OrderCreated"), conforme linguagem ubíqua do FTGO */
  type: string;
  /** Identificador do agregado que originou o evento */
  aggregateId: string;
  /** Momento em que o evento foi gerado */
  occurredAt: string;
  /** Dados específicos do evento */
  payload: TPayload;
}

export function createEvent<TPayload>(
  type: string,
  aggregateId: string,
  payload: TPayload
): DomainEvent<TPayload> {
  return {
    type,
    aggregateId,
    occurredAt: new Date().toISOString(),
    payload,
  };
}

File: services-java/servico-order/src/main/java/br/com/ftgo/order/domain/event/DomainEvent.java
package br.com.ftgo.order.domain.event;

import java.time.Instant;

import com.fasterxml.jackson.annotation.JsonProperty;

/**
 * Contrato genérico de evento de domínio publicado pelo ServicoOrder,
 * alinhado com a especificação de eventos da tabela "Especificação
 * final dos microsserviços do FTGO" (EC-M4.3).
 *
 * Em produção, cada evento seria publicado em um tópico Kafka
 * (ex.: {@code order_events}), conforme a EC-M5.2. Nesta primeira
 * versão (apenas o ServicoOrder), a publicação ocorre em um event bus
 * em memória (ver infrastructure.messaging.InMemoryEventPublisher),
 * substituível por um adaptador Kafka real quando os demais
 * microsserviços forem implementados.
 */
public interface DomainEvent {

    /** Nome do evento (ex.: "OrderCreated"), conforme linguagem ubíqua do FTGO. */
    @JsonProperty("type")
    String type();

    /** Identificador do agregado que originou o evento (orderId). */
    String aggregateId();

    /** Momento em que o evento foi gerado. */
    Instant occurredAt();
}
